backend/service/k8s/pods.go

```diff
@@ -62,6 +64,104 @@ func (s *svc) ListPods(ctx context.Context, clientset, cluster, namespace string
 	return pods, nil
 }
 
+// Update pod fields if the current field values match with that's described by expectedObjectMetaFields
+//
+// TODO: add support for updating pod labels
+func (s *svc) UpdatePod(ctx context.Context, clientset, cluster, namespace, name string, expectedObjectMetaFields *k8sapiv1.ExpectedObjectMetaFields, objectMetaFields *k8sapiv1.ObjectMetaFields, removeObjectMetaFields *k8sapiv1.RemoveObjectMetaFields) error {
+	if len(objectMetaFields.GetLabels()) > 0 || len(removeObjectMetaFields.GetLabels()) > 0 {
+		return errors.New("update of pod labels not implemented")
+	}
+
+	// Ensure that the caller is not trying to delete an annotation and update it at the same time
+	newAnnotations := objectMetaFields.GetAnnotations()
+	for _, annotation := range removeObjectMetaFields.GetAnnotations() {
+		_, annotationIsUpdated := newAnnotations[annotation]
+		if annotationIsUpdated {
+			return fmt.Errorf("annotation '%s' can't be updated and removed at once", annotation)
+		}
+	}
+
+	cs, err := s.manager.GetK8sClientset(clientset, cluster, namespace)
+	if err != nil {
+		return err
+	}
+
+	pod, err := cs.CoreV1().Pods(cs.Namespace()).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Check that the current state of the pod matches with expectedObjectMetaFields.
+	//
+	// If there is a mismatch, checkExpectedObjectMetaFields() will return an error with the list of mismatches.
+	err = checkExpectedObjectMetaFields(expectedObjectMetaFields, pod)
+	if err != nil {
+		return err
+	}
+
+	// Update/add annotations
+	podAnnotations := pod.GetAnnotations()
+	for annotation, value := range newAnnotations {
+		podAnnotations[annotation] = value
+	}
+
+	// Delete annotations to be removed
+	for _, annotation := range removeObjectMetaFields.GetAnnotations() {
+		delete(podAnnotations, annotation)
+	}
+
+	_, err = cs.CoreV1().Pods(cs.Namespace()).Update(pod)
+	return err
+}
+
+func checkExpectedObjectMetaFields(expectedObjectMetaFields *k8sapiv1.ExpectedObjectMetaFields, pod *corev1.Pod) error {
```

Comment:
This looks highly reusable why not just pass the resources `ObjectMeta` instead of a pod?